src/WebJobs.Script.WebHost/Management/LinuxInstanceManager.cs

```diff
@@ -0,0 +1,153 @@
+            }
+            else if (_assignmentContext == null)
+            {
+                lock (_assignmentLock)
+                {
+                    if (_assignmentContext != null)
+                    {
+                        return _assignmentContext.Equals(context);
+                    }
+                    _assignmentContext = context;
+                }
+
+                _logger.LogInformation($"Starting Assignment. Cloud Name: {_environment.GetCloudName()}");
+
+                // set a flag which will cause any incoming http requests to buffer
+                // until specialization is complete
+                // the host is guaranteed not to receive any requests until AFTER assign
+                // has been initiated, so setting this flag here is sufficient to ensure
+                // that any subsequent incoming requests while the assign is in progress
+                // will be delayed until complete
+                _webHostEnvironment.DelayRequests();
+
+                // start the specialization process in the background
+                Task.Run(async () => await Assign(context));
+
+                return true;
+            }
+            else
+            {
+                // No lock needed here since _assignmentContext is not null when we are here
+                return _assignmentContext.Equals(context);
+            }
+        }
+
+        public abstract Task<string> ValidateContext(HostAssignmentContext assignmentContext);
+
+        private async Task Assign(HostAssignmentContext assignmentContext)
+        {
+            try
+            {
+                // first make all environment and file system changes required for
+                // the host to be specialized
+                _logger.LogInformation("Applying {environmentCount} app setting(s)", assignmentContext.Environment.Count);
+                assignmentContext.ApplyAppSettings(_environment, _logger);
+                await ApplyContext(assignmentContext);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Assign failed");
+                await _meshServiceClient.NotifyHealthEvent(ContainerHealthEventType.Fatal, GetType(), "Assign failed");
+                throw;
+            }
+            finally
+            {
+                // all assignment settings/files have been applied so we can flip
+                // the switch now on specialization
+                // even if there are failures applying context above, we want to
+                // leave placeholder mode
+                _logger.LogInformation("Triggering specialization");
+                _webHostEnvironment.FlagAsSpecializedAndReady();
+
+                _webHostEnvironment.ResumeRequests();
+            }
+        }
+
+        protected abstract Task ApplyContext(HostAssignmentContext assignmentContext);
+
+        protected abstract Task<string> DownloadWarmupAsync(RunFromPackageContext context);
```

Comment:
small nit: "downloading" is an implementation detail of the atlas manager. Can we keep the name more generic?
```suggestion
        protected abstract Task<string> WarmupAsync(RunFromPackageContext context);
```